### PR TITLE
[installer] make sure dashboard is deployed after server and papi-server

### PR DESCRIPTION
**Pull request by mustard-mh**
## Description

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at 8857387</samp>

This pull request adds a feature flag mechanism to the service-waiter component using the ConfigCat SDK, and uses it to control the waiting for the public-api-server and the server components in the dashboard deployment. It also adds the component name as a custom attribute for experiments. The changes affect the `components/common-go/experiments`, `components/service-waiter/cmd`, and `install/installer/pkg/components/dashboard` packages.

</details>

## Related Issue(s)

Fixes EXP-822

## How to test

- Open PR with Gitpod
- Change anything in server i.e. add a console.log
- Exec `leeway dev:preview` pods should deployed successfully
- Change anything in public-api-server i.e. fmt.Println
- Exec `leeway dev:preview` pods should deployed successfully

**Control with Feature Flag**
- Change Feature Flag `service_waiter_skip_components` value to skip true
- It should skip all components' service-waiter
- With attribute `component` set with `server` or `public-api-server` or `server,public-api-server`, it should skip target service waiter

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - hw-exp-822</li>
	<li><b>🔗 URL</b> - <a href="https://hw-exp-822.preview.gitpod-dev.com/workspaces" target="_blank">hw-exp-822.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - hw-EXP-822-gha.19713</li>
	<li><b>🗒️ Logs</b> - <a href="https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-hw-exp-822%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-core-dev" target="_blank">GCP Logs Explorer</a></li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/service-waiter/cmd/component.go
Comment (akosyakov) [inline components/service-waiter/cmd/component.go:81]: should not it be false by default? otherwise if it fails to get configuration first it will continue?

Comment (mustard-mh) [reply components/service-waiter/cmd/component.go:81]: default true in case we can't stop service-waiter when we want to do it.

PR need to change since it needs configcat's env vars

I will check in preview env to see its behavior and see if first config get is slow

Comment (akosyakov) [reply components/service-waiter/cmd/component.go:81]: I don't think testing with preview envs will instil sense of security in me :) 

We should be sure that it is actual value from config cat. I don't believe there is a guarantee on it. There is a chance that we don't get a proper value and then we don't have a guarantee on deployment order.

Comment (akosyakov) [reply components/service-waiter/cmd/component.go:81]: cc @geropl for the opinion on it

File: components/service-waiter/cmd/component.go
Comment (geropl) [reply components/service-waiter/cmd/component.go:79]: I think this works.
In addition to the log.info in the caller, I'd add an log.error here: "failed to create experiments client".

File: components/service-waiter/cmd/component.go
Comment (geropl) [inline components/service-waiter/cmd/component.go:138]: Log for this specific case woudl be great, too!
Afterwards, we will always 100% understand which branch we took based on the log messages. :+1:

File: components/service-waiter/cmd/component.go
Comment (geropl) [reply components/service-waiter/cmd/component.go:81]: IMO we should default to a special value ("NONE"?): If we see this value, we re-fetch until we either a) hit a timeout, or b) until we get a valid value for the feature flag. On timeout, we should be loud about it in the logs. The timeout should be picked that it does not drastically extend the overall max. waiting time for a pod to be ready.

I think this is the only way to get it working with a feature flag reliably (if ConfigCat works, which we have to assume anyway), and still have an automatic fallback in case it does not. The feature flag being on only acts as a short-cut, then.

Comment (mustard-mh) [reply components/service-waiter/cmd/component.go:81]: If we want to get the actual value (wait if configcat fetch failed etc), we can simply change default to `false -> not skip` then retry logic will make it wait again.

If we want to make sure even configcat is broken we want a way to unblock ourselves, default to `true -> skip` works. (But I think this change will be quite stable (I have added unit test that it don't configure wrong target image 🙈🥹). So perberly we can ignore this case.)

A special value like `NONE` should behavior the same with default false.

So good if I change it to default false? @akosyakov @geropl

Comment (akosyakov) [reply components/service-waiter/cmd/component.go]: Agree with timeout. I wonder what is the reason of hiccups with configcat, proxy deployment?

File: components/service-waiter/cmd/component.go
Comment (akosyakov) [inline components/service-waiter/cmd/component.go:51]: Should not we call it each time inside the waitPodsImage loop?

Comment (mustard-mh) [reply components/service-waiter/cmd/component.go]: They behavior the same

File: components/service-waiter/cmd/component.go
Comment (mustard-mh) [inline components/service-waiter/cmd/component.go:151]: We need default client to nil, since it will never be nil without it -> it will return a client `NewAlwaysReturningDefaultValueClient`. Then `self-hosted` Gitpod Instance may need to keep wait for it

File: components/service-waiter/cmd/component.go
Comment (akosyakov) [inline components/service-waiter/cmd/component.go:141]: I would put rather high 10 mins

Comment (mustard-mh) [reply components/service-waiter/cmd/component.go:141]: How we know that feature flag proxy will stuck for more than 10 minutes?

**Review by akosyakov — APPROVED** [review]
We discussed that to land it the metric should be added. We deploy it with false state, then we enable it on dogfood, then on Cloud, and afterwards for other cells.

pre-approved since we talked about it, but i did not try and metric should be added

/hold

**Comment by mustard-mh** [comment]
@akosyakov Force push changes into this commit https://github.com/gitpod-io/gitpod/pull/19042/commits/470749f78a6ce2cb4e79a5fca99cd11d3793c3e3

Could you take a look? It may happen in preview env that pods ready too fast so there's no related metrics for them

**Comment by mustard-mh** [comment]
Checked with preview env, that metric works. It took me a lot of time to debug with service-waiter. Configcat config fetch seems will failed for the start.

🥲 And I just recognized that we forget the very first idea to wait: if not skip, we need to keep fetch feature flag. In case it the value is changed.

Addressed with https://github.com/gitpod-io/gitpod/pull/19042/commits/198d71d79326ed1a46e2bb225e2f656c2f35ef20 , metric test result see below 

```
/ # curl http://localhost:9500/metrics | grep service_waiter
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  3690    0  3690    0     0  3673k      0 --:--:-- --:--:-- --:--:-- 3603k
# HELP service_waiter_skip_components_result_total Total number of wait result of service_waiter/component service_waiter_skip_components flag
# TYPE service_waiter_skip_components_result_total counter
service_waiter_skip_components_result_total{ok="true",value="false"} 3
/ # 
```

**Comment by mustard-mh** [comment]
Tested again with deploy server / public-api with `leeway run dev:preview` we are good to go if metric is alright @akosyakov

**Comment by akosyakov** [comment]
@geropl Could you help reviewing it?

**Comment by akosyakov** [comment]
> Could you take a look? It may happen in preview env that pods ready too fast so there's no related metrics for them

Do you mean first deployment or each time?

**Comment by geropl** [comment]
Will test now, too!

**Comment by mustard-mh** [comment]
> > Could you take a look? It may happen in preview env that pods ready too fast so there's no related metrics for them
> 
> Do you mean first deployment or each time?

@akosyakov  It'll not affect component deployment, but we don't have metric in this case, since feature flag's value is not got yet.

File: components/service-waiter/cmd/component.go
Comment (geropl) [inline components/service-waiter/cmd/component.go:192]: :orange_heart:

File: components/service-waiter/cmd/component_test.go
Comment (geropl) [inline components/service-waiter/cmd/component_test.go:40]: :orange_heart:

File: components/service-waiter/cmd/component.go
Comment (geropl) [inline components/service-waiter/cmd/component.go:168]: @mustard-mh When waiting for longer than timeout, I got a integer divide by zero here (I scaled server down to 0 replicas to test):
```
panic: runtime error: integer divide by zero

goroutine 18 [running]:
github.com/gitpod-io/gitpod/service-waiter/cmd.startWaitFeatureFlag.func1()
        github.com/gitpod-io/gitpod/service-waiter/cmd/component.go:168 +0x4e5
github.com/gitpod-io/gitpod/service-waiter/cmd.startWaitFeatureFlag({0x1a40810?, 0xc000434c40?}, 0x0?)
        github.com/gitpod-io/gitpod/service-waiter/cmd/component.go:174 +0x1be
created by github.com/gitpod-io/gitpod/service-waiter/cmd.glob..func2 in goroutine 1
        github.com/gitpod-io/gitpod/service-waiter/cmd/component.go:56 +0x45b
```

**Comment by geropl** [comment]
Finished testing:
 - happy path: :heavy_check_mark: 
 - skip with feature flag: :heavy_check_mark: 
 - skip after timeout: got an error (see [comment](https://github.com/gitpod-io/gitpod/pull/19042#discussion_r1392376117))

I double checked the code and tests, looks really good! :pray: Once the last thing is fixed we should :checkered_flag: and merge. :rocket:

**Comment by mustard-mh** [comment]
Addressed, thank you 🚢

**Pull request merged**
